packages/graphql-elasticsearch-transformer/streaming-lambda/es-external-version/python_streaming_function.py

```diff
@@ -0,0 +1,249 @@
+        if record.get('eventSource') == 'aws:dynamodb':
+            ddb = record['dynamodb']
+            ddb_table_name = get_table_name_from_arn(record['eventSourceARN'])
+            doc_seq = ddb['SequenceNumber']
+        elif record.get('eventSource') == 'aws:kinesis':
+            ddb = json.loads(base64.b64decode(record['kinesis']['data']))
+            ddb_table_name = ddb['SourceTable']
+            doc_seq = record['kinesis']['sequenceNumber']
+        else:
+            logger.error('Ignoring non-DynamoDB event sources: %s',
+                        record.get('eventSource'))
+            continue
+
+        # Compute DynamoDB table, type and index for item
+        doc_table = ddb_table_name.lower()
+        doc_type = DOC_TYPE
+        doc_table_parts = doc_table.split('-')
+        doc_es_index_name = doc_table_parts[0] if len(doc_table_parts) > 0  else doc_table
+
+        # Dispatch according to event TYPE
+        event_name = record['eventName'].upper()  # INSERT, MODIFY, REMOVE
+        logger.debug('doc_table=%s, event_name=%s, seq=%s',
+                    doc_table, event_name, doc_seq)
+
+        # Treat events from a Kinesis stream as INSERTs
+        if event_name == 'AWS:KINESIS:RECORD':
+            event_name = 'INSERT'
+
+        # Update counters
+        if event_name == 'INSERT':
+            cnt_insert += 1
+        elif event_name == 'MODIFY':
+            cnt_modify += 1
+        elif event_name == 'REMOVE':
+            cnt_remove += 1
+        else:
+            logger.warning('Unsupported event_name: %s', event_name)
+
+        is_ddb_insert_or_update = (event_name == 'INSERT') or (event_name == 'MODIFY')
+        is_ddb_delete = event_name == 'REMOVE'
+        image_name = 'NewImage' if is_ddb_insert_or_update else 'OldImage'
+
+        if image_name not in ddb:
+            logger.warning(
+                'Cannot process stream if it does not contain ' + image_name)
+            continue
+        logger.debug(image_name + ': %s', ddb[image_name])
+        # Deserialize DynamoDB type to Python types
+        doc_fields = ddb_deserializer.deserialize({'M': ddb[image_name]})
+
+        logger.debug('Deserialized doc_fields: %s', doc_fields)
+
+        if ('Keys' in ddb):
+            doc_id = compute_doc_index(ddb['Keys'], ddb_deserializer)
+        else:
+            logger.error('Cannot find keys in ddb record')
+
+        # Generate JSON payload
+
+        # If DynamoDB INSERT or MODIFY, send 'index' to ES
+        if is_ddb_insert_or_update:
+            # Generate ES payload for item
+            action = {'index': {'_index': doc_es_index_name,
+                                'version_type': 'external',
+                                '_version': doc_fields['_version'],
+                                '_type': doc_type, '_id': doc_id}}
+            # Action line with 'index' directive
+            es_actions.append(json.dumps(action))
+            del doc_fields['_version']
```

Comment:
Why does this get deleted if you're using external versions?

Reply:
It might not be necessary to delete when using external versioning, but it's redundant. However, not deleting it may mean we don't have to modify the resolvers to merge the ES version with the record. I'll try it out and if that works then I'll leave it in

Reply:
Looks like leaving it in, even with external versioning on still causes conflicts:
```
{
    "index": {
        "_index": "message",
        "_type": "doc",
        "_id": "9aa4eca1-6978-406a-b0eb-ecc76aaa49bd",
        "status": 400,
        "error": {
            "type": "mapper_parsing_exception",
            "reason": "Cannot generate dynamic mappings of type [_version] for [_version]"
        }
    }
}
```